[DOCS] Refactored index-templates topic. (#59737)
elastic/elasticsearch: docs/reference/indices/put-index-template.asciidoc

[[indices-put-template]]
=== Put index template API
++++
<titleabbrev>Put index template</titleabbrev>
++++

Creates or updates an index template.
Index templates define <<index-modules-settings,settings>>, <<mapping,mappings>>, 
and <<indices-aliases,aliases>> that can be applied automatically to new indices. 

[source,console]
--------------------------------------------------
PUT /_index_template/template_1
{
  "index_patterns" : ["te*"],
  "priority" : 1,
  "template": {
    "settings" : {
      "number_of_shards" : 2
    }
  }
}
--------------------------------------------------
// TESTSETUP

//////////////////////////

[source,console]
--------------------------------------------------
DELETE _index_template/template_*
--------------------------------------------------
// TEARDOWN

//////////////////////////

[[put-index-template-api-request]]
==== {api-request-title}

`PUT /_index_template/<index-template>`

[[put-index-template-api-desc]]
==== {api-description-title}

{es} applies templates to new indices based on an

> index pattern that matches the index name.

Reviewer: I'd use `wildcard pattern` instead of `index pattern`.

That pattern can now match the name of a data stream.

This number is not automatically generated by {es}.

`_meta`::
(Optional, object)
Optional user metadata about the index template. May have any contents.
This map is not automatically generated by {es}.

[[put-index-template-api-example]]
==== {api-examples-title}

===== Index template with index aliases

You can include <<indices-aliases,index aliases>> in an index template.

[source,console]
--------------------------------------------------
PUT _index_template/template_1
{
  "index_patterns" : ["te*"],
  "template": {
    "settings" : {
        "number_of_shards" : 1
    },
    "aliases" : {
        "alias1" : {},
        "alias2" : {
            "filter" : {
                "term" : {"user" : "kimchy" }
            },
            "routing" : "kimchy"
        },
        "{index}-alias" : {} <1>
    }
  }
}
--------------------------------------------------
<1> the `{index}` placeholder in the alias name will be replaced with the
actual index name that the template gets applied to, during index creation.

[[multiple-templates]]
===== Indices matching multiple templates

When an index is created that matches multiple index templates, only the index template with the

> highest priority is applied. For example:

Reviewer: I think this applies to data streams as well. I'd update to indicate that.

NOTE: Multiple templates with overlapping index patterns at the same priority are not allowed, and
an error will be thrown when attempting to create a template matching an existing index template at
identical priorities.

[[versioning-templates]]
===== Template versioning

You can use the `version` parameter to add an optional version number to an index template. External
systems can use these version numbers to simplify template management.

The `version` parameter is completely optional and not automatically generated by {es}.

To unset a `version`, replace the template without specifying one.

[source,console]
--------------------------------------------------
PUT /_index_template/template_1
{
  "index_patterns" : ["foo", "bar"],
  "priority" : 0,
  "template": {
    "settings" : {
        "number_of_shards" : 1
    }
  },
  "version": 123
}
--------------------------------------------------

To check the `version`, you can use the <<indices-get-template, get index template>> API.

[[template-metadata]]
===== Template metadata

You can use the `_meta` parameter to add optional metadata to an index template. This is a

> user-defined map that can contain any data. This data will be stored in the cluster state however,

Reviewer: nit
```suggestion
user-defined object that can contain any data. This data will be stored in the cluster state however,
```

so keeping it short is preferrable.

The `_meta` parameter is completely optional and not automatically generated by {es}.

To unset `_meta`, replace the template without specifying one.

[source,console]
--------------------------------------------------
PUT /_index_template/template_1
{
  "index_patterns": ["foo", "bar"],
  "template": {
    "settings" : {
        "number_of_shards" : 3
    }
  },
  "_meta": {
    "description": "set number of shards to three",
    "serialization": {
      "class": "MyIndexTemplate",
      "id": 17
    }
  }
}
--------------------------------------------------

To check the `_meta`, you can use the <<indices-get-template, get index template>> API.

[[data-stream-definition]]
===== Data stream definition

> If an index template should create data streams, the template must include an

Reviewer: ```suggestion
If an index template is used for data streams, the template must include an
```

> empty `data_stream` object. See <<create-a-data-stream-template>>.

Reviewer: It may be helpful to mention that these index templates cannot be used for regular indices.
